adding game mechanics apis

yorozuya-2003 opened the pull request.

yorozuya-2003 committed:
added api functions for game mechanics (board state, dice, move)

yorozuya-2003 committed:
applied transaction to game and user operations; fixed minor bugs in the game logic

manishsuperhr commented on backend/controllers/gameController.js:
```diff
@@ -9,15 +14,15 @@ const create_game = async (req, res) => {
     await client.query("BEGIN");
     const availableGames = await client.query(
-      "SELECT * FROM ludo.game WHERE status='WAITING_FOR_PLAYERS' ORDER BY created_on DESC"
+      "SELECT * FROM game WHERE status='WAITING_FOR_PLAYERS' ORDER BY created_on DESC"
```
SELECT * FROM game WHERE status='WAITING_FOR_PLAYERS' ORDER BY created_on DESC LIMIT 1

manishsuperhr replied:
PUT an INDEX on status

manishsuperhr commented on backend/controllers/gameController.js:
```diff
@@ -89,4 +93,255 @@ const create_game = async (req, res) => {
+  try {
+    await client.query("BEGIN");
+    const gameId = parseInt(req.headers.game_id);
+    const coinStates = await client.query(
```
Move this block of code after checking "if game has ended" This way you will save on this query exdcution on database !

manishsuperhr commented on backend/controllers/gameController.js:
```diff
@@ -89,4 +93,255 @@ const create_game = async (req, res) => {
+    await client.query("COMMIT");
+    res
+      .status(200)
+      .json({ board_state: coinStateData, game_has_ended: gameHasEnded });
```
If gameHasEnded, send board_state as Null !

manishsuperhr commented on backend/controllers/gameController.js:
```diff
@@ -89,4 +93,255 @@ const create_game = async (req, res) => {
+
+  try {
+    await client.query("BEGIN");
+    const gameId = req.headers.game_id;
```
We can detect the user_id based on access_token/request_header/request_body, get player id using the user_id in player table.

Now check if the turn belongs to the same player or not !

manishsuperhr commented on backend/controllers/gameController.js:
```diff
@@ -89,4 +93,255 @@ const create_game = async (req, res) => {
+      [gameId]
+    );
+
+    const gameHasEnded = playerTurn.rows[0].count === 0;
```
The correct way to check if a game has ended if to check game.status==="GAME_ENDED"

manishsuperhr commented on backend/controllers/gameController.js:
```diff
@@ -89,4 +93,255 @@ const create_game = async (req, res) => {
+    const gameHasEnded = playerTurn.rows[0].count === 0;
+
+    // if game has not ended
+    if (!gameHasEnded) {
```
defensive programming.

check first for if game has ended.
No need for else block in that case !

manishsuperhr commented on backend/controllers/gameController.js:
```diff
@@ -89,4 +93,255 @@ const create_game = async (req, res) => {
+      let moveableCoins;
+      if (diceValue === 6) {
+        moveableCoins = await client.query(
+          "SELECT COUNT(id) FROM coin_state WHERE player_id = $1 AND position <= $2",
```
eliminate the coins in this query whose journey is complete

manishsuperhr commented on backend/controllers/gameController.js:
```diff
@@ -89,4 +93,255 @@ const create_game = async (req, res) => {
+          [playerId, value]
+        );
+      } else {
+        moveableCoins = await client.query(
```
eliminate the coins in this query whose journey is complete

manishsuperhr commented on backend/controllers/gameController.js:
```diff
@@ -89,4 +93,255 @@ const create_game = async (req, res) => {
+  try {
+    await client.query("BEGIN");
+
+    const coinId = req.headers.coin_id;
```
I think we can take game_id and coin_id in request path/body.

Derive everything else from DB.

manishsuperhr commented on backend/controllers/gameController.js:
```diff
@@ -89,4 +93,255 @@ const create_game = async (req, res) => {
+    const coinId = req.headers.coin_id;
+    const coinColor = req.headers.coin_color;
+    const playerId = req.headers.player_id;
+    const gameId = req.headers.game_id;
```
If you don't ask for values that you can infer from DB, you won't have to validate if the user sent the correct data or not !

manishsuperhr commented on backend/controllers/gameController.js:
```diff
@@ -89,4 +93,255 @@ const create_game = async (req, res) => {
+
+      // getting unfinished players (as well as current player in order)
+      const unfinishedPlayers = await client.query(
+        "SELECT * FROM player WHERE (game_id = $1 AND status = 'IN_GAME') OR player_id = $2 ORDER BY id",
```
SELECT id, other columns that are required... rather than selecting *

SELECT id FROM player WHERE game_id = $1 AND status = 'IN_GAME';; It returns 4 rows with 1 column only. do the math in application layer.

yorozuya-2003 committed:
modified game mechanics apis; added dice to player_turn table
- added dice to player_turn table
- added index on status in game table
- added game rankings controller and route
- added LIMIT to SELECT queries
- added checks for game start / exist / finish
- added checks for matching request header and db values

yorozuya-2003 commented:
### Intuition behind _dice_ in _player_turn_ table

**Roll dice**
_If (dice != 0) previous player did not move their coin yet_
- If any coin movable:
    Dice updated to dice value generated
- Else if no movable coin:
    Update player turn to next player and make Dice = 0 in player_turn table


**Move coin**
_If (dice == 0) current player did not roll dice yet_
- Coin moved
- Update player turn to next player and make Dice = 0 in player_turn table

yorozuya-2003 committed:
added player-turn id in game-state api; check for valid coin-id in move-coin api
- modified status codes to 400 in case of mismatch between request headers and db data
- removed dice value from required request headers in move-coin api
- removed unnecessary LIMIT 1 statement from player_turn table select queries

yorozuya-2003 committed:
added new dice value logic and fixed bugs: due to parseInt and absence of return; statements
and other bugs related to game logic

yorozuya-2003 committed:
modified cut condition
cut will now set position of all coins (instead of only 1) on that position to -1

yorozuya-2003 committed:
added cors to backend
